### PR TITLE
feat: add identify protocol example

### PR DESCRIPTION
**Pull request by acul71**
The Identify Protocol example was missing in `py-libp2p`, making it difficult for users to test and understand how the identify protocol works in a real peer-to-peer scenario.  

## How was it fixed?  

Added a standalone `identify.py` example demonstrating the Identify Protocol in action. The example sets up two hosts:  

- **Host A (Listener)**: Waits for an incoming identify request.  
- **Host B (Dialer)**: Connects to Host A and initiates the identify protocol exchange.  

The output includes details such as:  

- The remote peer's observed address  
- Supported protocols  
- Listen addresses  
- Public key information  

```bash
python identify.py 
First host listening. Run this from another console:

python identify.py -p 8889 -d /ip4/0.0.0.0/tcp/8888/p2p/QmPF97e2nktakZVBzmn1uDGGZTaDyDLDtVR8MLLrH6UmYz

Waiting for incoming identify request...

```

```bash
python identify.py -p 8889 -d /ip4/0.0.0.0/tcp/8888/p2p/QmPF97e2nktakZVBzmn1uDGGZTaDyDLDtVR8MLLrH6UmYz
dialer (host_b) listening on /ip4/0.0.0.0/tcp/8889
Second host connecting to peer: QmPF97e2nktakZVBzmn1uDGGZTaDyDLDtVR8MLLrH6UmYz
Starting identify protocol...
Identify response:
  Public Key (Base64): CAASpgIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCSPdishWS5vDq5H1x/T2Za8Sp6pQ10bwlyfpe0BdoNBtCAkVvVtR2EiSa7bQYVAMK6QH7Hzxdxs/LR4rwaK20fS9PLQ0I7v4Pkk+wCAne2EGSDrrXGRRC9tNaWWNujaRlNSmWlQH3V/W8Y2MGJIl6uezwKt3EUZQ12SMDgeTFGogvNnf5KYwlD3GczpButKgBkeeerscBZzJvm+JGIDqWBGZF4BPye691EGH8Nj2XMKm4Adj5Yew9UnSxI1mNIFksNEzhuN6j1NrzWjnL3cgIfUaYBS7Or/zatcayY+WiwQSZzcpVB0lIp06NLrbIx3GWdV6Y9LfZ9AlKfMMYG4ZTRAgMBAAE=
  Listen Addresses: ['/ip4/0.0.0.0/tcp/8888/p2p/QmPF97e2nktakZVBzmn1uDGGZTaDyDLDtVR8MLLrH6UmYz']
  Protocols: ['/ipfs/id/1.0.0', '/ipfs/ping/1.0.0']
  Observed Address: ['/ip4/127.0.0.1/tcp/37498']
  Protocol Version: ipfs/0.1.0
  Agent Version: py-libp2p/0.2.0

```

### To-Do  

- [ ] Clean up commit history  
- [ ] Add or update documentation related to these changes  
- [ ] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)  

#### Cute Animal Picture  

Cute Animal

### CODE REVIEW

File: examples/identify/identify.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+            await host_b.connect(info)
+            stream = await host_b.new_stream(info.peer_id, (IDENTIFY_PROTOCOL_ID,))
+
+            try:
+                print("Starting identify protocol...")
+                response = await stream.read()
+                await stream.close()
+                identify_msg = Identify()
+                # print(f"Response: {response}")
```
Comment (pacrob) [inline examples/identify/identify.py:90]: Should this be removed?

File: tests/core/identity/identify/test_protocol.py
Comment (pacrob) [inline tests/core/identity/identify/test_protocol.py:1]: It seems like naming the file `test_identify.py` would be clearer. Is there a reason to keep it `test_protocol.py`?

File: libp2p/identity/identify/protocol.py
Comment (pacrob) [inline libp2p/identity/identify/protocol.py:1]: Ah, probably because this is `protocol.py`. Again, it seems clearer to have this be `identify.py`.

File: libp2p/identity/identify/protocol.py [hunk 2]
```diff
@@ -15,44 +18,89 @@
...
+    get_agent_version,
+)
 
 from .pb.identify_pb2 import (
     Identify,
 )
 
+# Not sure I can do this or I break a pattern
+# logger = logging.getLogger("libp2p.identity.identify")
+logger = logging.getLogger(__name__)
```
Comment (pacrob) [inline libp2p/identity/identify/protocol.py:31]: From the rest of the lib, it appears that 

```python
logger = logging.getLogger("libp2p.identity.identify.protocol")
```

is what would work.

**Comment by pacrob** [comment]
Looking great, @acul71!

 - You've added the `get_remote_address` function across a number of libp2p modules, but I don't see it being tested anywhere. Ideally we'd have at least a basic test for each.
 - You can add the demo to the list in `setup.py`, down at the bottom within `console_scripts`.

I've left a few notes throughout the code. Let me know what you think.

File: examples/identify/identify.py (hunk 1 above)
Comment (acul71) [reply examples/identify/identify.py:90]: Yep forgot...

File: tests/core/identity/identify/test_protocol.py
Comment (acul71) [reply tests/core/identity/identify/test_protocol.py:1]: Yes

File: libp2p/identity/identify/protocol.py (hunk 2 above)
Comment (acul71) [reply libp2p/identity/identify/protocol.py:31]: The logger name was without `.protocol` before. 
I'll use logging.getLogger(__name__) so don't have to bother with file name change.
```logging.getLogger(__name__) == logging.getLogger("libp2p.identity.identify.protocol")```

File: libp2p/identity/identify/protocol.py
Comment (acul71) [reply libp2p/identity/identify/protocol.py:1]: renamed protocol.py to identify.py

**Comment by acul71** [comment]
> Looking great, @acul71!

> * You've added the `get_remote_address` function across a number of libp2p modules, but I don't see it being tested anywhere. Ideally we'd have at least a basic test for each.

TODO: Planning to add more tests 

> * You can add the demo to the list in `setup.py`, down at the bottom within `console_scripts`.

DONE

> I've left a few notes throughout the code. Let me know what you think.

Changed name form protocol.py to identify.py as you suggested, removed comments.

**Comment by seetadev** [comment]
> > Looking great, @acul71!
> 
> > * You've added the `get_remote_address` function across a number of libp2p modules, but I don't see it being tested anywhere. Ideally we'd have at least a basic test for each.
> 
> TODO: Planning to add more tests
> 
> > * You can add the demo to the list in `setup.py`, down at the bottom within `console_scripts`.
> 
> DONE
> 
> > I've left a few notes throughout the code. Let me know what you think.
> 
> Changed name form protocol.py to identify.py as you suggested, removed comments.

@acul71 @pacrob : This looks great. We will need to add the identify example along with the other 3 examples of py-libp2p.

@pacrob : Thank you so much for your valuable guidance and helping us arrive at a good conclusion on Identify protocol example. The feedback shared is very helpful and will enable us to bootstrap other examples.

Our longer term goal is to have atleast all the features and examples that we see at rust-libp2p(please visit https://github.com/libp2p/rust-libp2p/tree/master/examples ) and js-libp2p (please visit https://github.com/libp2p/js-libp2p-examples/tree/main/examples ).

The feedback shared in this issue will be used in developing the other py-libp2p features and examples.

@acul71 : Wish if you could also share a document adding @pacrob 's feedback and pointers (discussed in this pull request) for new devs interested in developing key examples in py-libp2p as shared above.

**Comment by acul71** [comment]
# py-libp2p Developer Guide: Feedback & Best Practices for Examples

This document compiles key feedback and pointers from recent code reviews and discussions. It is intended to serve as a reference for new contributors working on examples and core features for py-libp2p. Our goal is to align with the quality and functionality seen in rust-libp2p and js-libp2p examples.

## 1. Overview of py-libp2p Examples

py-libp2p aims to provide a Python implementation of libp2p protocols and features. Our longer-term goal is to match the breadth of features found in [rust-libp2p examples](https://github.com/libp2p/rust-libp2p/tree/master/examples) and [js-libp2p examples](https://github.com/libp2p/js-libp2p-examples/tree/main/examples). The examples are not only demos but also act as living documentation for developers exploring protocol interactions.

## 2. Key Feedback from Recent Reviews

### a. Testing and Quality Assurance
- **get_remote_address Function:**  
  - The `get_remote_address` function has been added across multiple modules. However, it currently lacks dedicated tests.
  - **Action Item:** Add at least basic tests for every module that uses this function. This will help ensure that any future changes do not inadvertently break its functionality.

### b. Code Organization and Naming Conventions
- **File Renaming:**  
  - The file previously named `protocol.py` has been renamed to `identify.py` to better reflect its purpose.
- **Comments and Clean Code:**  
  - Unnecessary comments have been removed to clean up the codebase.
  - **Action Item:** When making further changes, ensure that any remaining inline comments are clear and add value without cluttering the code.

### c. Examples and Demo Integration
- **Adding the Identify Protocol Example:**  
  - In addition to the three existing examples, there is now a need to include a dedicated Identify protocol example.
  - **Action Item:** Develop a comprehensive example for the Identify protocol and add it alongside the other demos.
- **Demo Registration in setup.py:**  
  - The demo has been added to the list in `setup.py` under `console_scripts`.
  - **Action Item:** Verify that the demo registration works correctly when installed via pip.

## 3. Best Practices for Developing Examples

### a. Consistency with Rust and JS Examples
- **Reference Implementations:**  
  - Use rust-libp2p and js-libp2p examples as references for structure, feature set, and usability.
  - Ensure that the examples in py-libp2p are self-contained, well-documented, and serve as learning tools for protocol interactions.

### b. Testing and Documentation
- **Write Tests:**  
  - Every new feature or protocol example should be accompanied by tests. Start with basic tests to validate the core functionality and then expand to cover edge cases.
- **Document the Code:**  
  - Maintain clear documentation within the code and in supplementary developer guides.
  - Use docstrings and inline comments sparingly, ensuring they add clarity.

### c. Feedback Incorporation Process
- **Iterative Improvements:**  
  - Use feedback from code reviews (such as those provided by @pacrob) to guide further improvements.
  - Encourage discussion within the community for any proposed changes, ensuring that solutions are robust and future-proof.
- **Sharing Knowledge:**  
  - This document itself is a living reference. Update it as new feedback or guidelines emerge to help onboard future contributors.

## 4. Next Steps for Contributors

1. **Implement Testing:**  
   - Prioritize writing tests for the `get_remote_address` function across all relevant modules.
2. **Develop and Add Examples:**  
   - Complete the Identify protocol example and ensure it is consistent with other demos.
3. **Update Documentation:**  
   - Use this guide to inform any new documentation you produce. If you have further insights or improvements, consider updating this document.
4. **Continuous Feedback:**  
   - Engage with the team (including reviewing PRs and providing feedback like @pacrob) to keep improving the codebase and examples.

## 5. Additional Resources

- [rust-libp2p Examples](https://github.com/libp2p/rust-libp2p/tree/master/examples)
- [js-libp2p Examples](https://github.com/libp2p/js-libp2p-examples/tree/main/examples)
- py-libp2p Repository: [GitHub - libp2p/py-libp2p](https://github.com/libp2p/py-libp2p)

---

This guide should help streamline contributions and ensure consistency across the project. Please refer to it whenever you work on new examples or core features in py-libp2p. Let’s continue to build a robust and developer-friendly libp2p ecosystem in Python!

**Comment by acul71** [comment]
I will add a video in the next days

**Comment by acul71** [comment]
@pacrob @seetadev 
# Test Analysis: Remote Address Delegation Chain in libp2p

## Overview
This test verifies the proper delegation of remote addresses through the entire libp2p network stack, ensuring address consistency across all protocol layers.

## Test Description
`test_complete_remote_address_delegation_chain` establishes a connection between two libp2p hosts and examines how the remote address is passed through each layer of the networking stack.

## Key Components Tested
- NetStream
- MplexStream
- Mplex
- SecureSession
- RawConnection
- TrioTCPStream
- Socket

## Test Flow
1. Creates two libp2p hosts (host_a and host_b)
2. Establishes a stream from host_a to host_b using the identify protocol
3. Retrieves references to components throughout the connection stack
4. Extracts the remote address from each layer
5. Verifies address consistency across all layers
6. Confirms the retrieved address matches host_b's actual address

## Significance
This test ensures the address delegation system functions correctly, which is essential for proper peer identification, connection management, and network reliability in the libp2p ecosystem.

**Comment by acul71** [comment]
@pacrob 
Hi Paul, this is a recap for making an example:

To add an example (ex: identify) :
- Create a directory in `examples/identify` 
- Create a file `examples/identify/identify.py` with the code of the example ( don't forget `__init__.py` making it a proper Python package so that is automatically discovered by find_packages() in `setup.py` )
- Add the example in the example list `docs/examples.rst`
- Add the example documentation in `docs/examples.identify.rst`
- Add a news fragment for the new release in file `newsfragments/536.feature.rst` (fix-id.type.rst)
- Generate doc files with `make docs` or `make linux-docs` in linux ( it generates files `libp2p.identity.identify.rst libp2p.identity.rst libp2p.identity.identify.pb.rst` )
- Add the example to `setup.py` 
  ```python
    entry_points={
        "console_scripts": [
            "chat-demo=examples.chat.chat:main",
            "echo-demo=examples.echo.echo:main",
            "ping-demo=examples.ping.ping:main",
            "identify-demo=examples.identify.identify:main",
        ],
    },
  ```
- do `make package-test` to test the release

  ```bash
   .....
   Activate with `source /tmp/tmpb9ybjgtg/package-smoke-test/bin/activate`
   Press enter when the test has completed. The directory will be deleted.
   ```
   test the example 
   ```bash
   source /tmp/tmpb9ybjgtg/package-smoke-test/bin/activate
   (package-smoke-test) $ identify-demo
   ```

**Comment by acul71** [comment]
@pacrob 
Hi Paul I did some fixing about the example in this PR:
https://github.com/libp2p/py-libp2p/pull/544
I tried to add https://github.com/libp2p/py-libp2p/pull/544/commits/9a18a61651e7ac57cfeee267ee58952138c51c15 (the actual changes here but I could not not sure why. )
